### PR TITLE
Dynamic cert file ca bundle

### PR DESCRIPTION
**Pull request by deads2k**
Dynamically reloads client certificates for both authentication and serving hints.

```release-note
client-ca bundles for the all generic-apiserver based servers will dynamically reload from disk on content changes
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/server/config.go [hunk 1]
```diff
@@ -346,21 +346,19 @@ func DefaultOpenAPIConfig(getDefinitions openapicommon.GetOpenAPIDefinitions, de
...
-		}
+func (c *AuthenticationInfo) ApplyClientCert(clientCA dynamiccertificates.CAContentProvider, servingInfo *SecureServingInfo) error {
+	if servingInfo == nil {
+		return nil
+	}
+	if clientCA == nil {
+		return nil
+	}
+	if servingInfo.ClientCA == nil {
+		servingInfo.ClientCA = clientCA
```
Comment (jackkleeman) [inline staging/src/k8s.io/apiserver/pkg/server/config.go]: Another option would be to get `NewUnionCAContentProvider` to simply return value A if value B is nil

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/dynamicfile_content.go [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+	defer c.queue.ShutDown()
+
+	klog.Infof("Starting %s", c.name)
+	defer klog.Infof("Shutting down %s", c.name)
+
+	// doesn't matter what workers say, only start one.
+	go wait.Until(c.runWorker, time.Second, stopCh)
+
+	// start timer that rechecks every minute, just in case.  this also serves to prime the controller quickly.
+	_ = wait.PollImmediateUntil(1*time.Minute, func() (bool, error) {
```
Comment (jackkleeman) [inline staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/dynamicfile_content.go]: If we are going to poll inside of the dynamic providers, do we need to poll in the main controller in tlsconfig.go?

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/dynamicfile_content.go [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+	caBundle, err := ioutil.ReadFile(c.filename)
+	if err != nil {
+		return err
+	}
+	if len(caBundle) == 0 {
+		return fmt.Errorf("missing content for CA bundle %q", c.Name())
+	}
+
+	// check to see if we have a change. If the values are the same, do nothing.
+	existing, ok := c.caBundle.Load().(*caBundleAndVerifier)
```
Comment (jackkleeman) [inline staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/dynamicfile_content.go]: It's worth noting that we store pointers in the `Value` here, but store a non-pointer `tls.Config`; is there a particular reason either way?

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/dynamicfile_content.go [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+	if ok && existing != nil && reflect.DeepEqual(existing.caBundle, caBundle) {
+		return nil
+	}
+
+	caBundleAndVerifier, err := newCABundleAndVerifier(c.Name(), caBundle)
+	if err != nil {
+		return err
+	}
+	c.caBundle.Store(caBundleAndVerifier)
```
Comment (jackkleeman) [inline staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/dynamicfile_content.go]: It seems to me that we cache the overall tls.Config in an atomic.Value, and cache the consituent dynamic elements also. Why not simply have `CurrentCABundleContent` always do a read off of disk, and have the only stateful element be the `caBundleContent`

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/dynamicfile_content.go [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+
+	for _, listener := range c.listeners {
+		listener.Enqueue()
```
Comment (jackkleeman) [inline staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/dynamicfile_content.go]: It looks like we cause an enqueue irrelevant of whether the bundle changes

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/dynamicfile_content.go (hunk 4 above)
Comment (jackkleeman) [reply staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/dynamicfile_content.go]: I feel like if we have 10 SNI certs using a similar approach, thats a lot of controllers polling, or we would need a separate SNIProvider that could unify them. If we always read the file on `CurrentCABundleContent`, then really we just have one goroutine polling, and maybe some file watches too

I feel like it makes sense for a `Provider` to know how to build a completely fresh certificate; and callers cache it as desired

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticatorfactory/requestheader.go [hunk 6]
```diff
@@ -28,9 +28,19 @@ type RequestHeaderConfig struct {
...
-	VerifyOptionFn x509request.VerifyOptionFunc
+	// CAContentProvider the options for verifying incoming connections using mTLS.  Generally this points to CA bundle file which is used verify the identity of the front proxy.
+	//	It may produce different options at will.
+	CAContentProvider CAContentProvider
 	// AllowedClientNames is a list of common names that may be presented by the authenticating front proxy.  Empty means: accept any.
 	AllowedClientNames []string
 }
+
+// CAContentProvider provides ca bundle byte content
+type CAContentProvider interface {
```
Comment (jackkleeman) [inline staging/src/k8s.io/apiserver/pkg/authentication/authenticatorfactory/requestheader.go]: How come this needs to be duplicated? Import cycle?

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/dynamicfile_content.go (hunk 2 above)
Comment (deads2k) [reply staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/dynamicfile_content.go]: > If we are going to poll inside of the dynamic providers, do we need to poll in the main controller in tlsconfig.go?

I think we should.  They control two different things.  The individual files can trigger the overall, but we also want a backstop on these.  It's more about implementations of the interface you did not create.

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/dynamicfile_content.go (hunk 3 above)
Comment (deads2k) [reply staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/dynamicfile_content.go]: > It's worth noting that we store pointers in the `Value` here, but store a non-pointer `tls.Config`; is there a particular reason either way?

This felt more natural.  I don't have a strong opinion.  I guess the pointer for tlsconfig may be more pure. If it shows up in a profile we can adjust.

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/dynamicfile_content.go (hunk 4 above)
Comment (deads2k) [reply staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/dynamicfile_content.go]: Files are only one implementation.  I'm actually driving this towards consuming a list/watch from the API.  It's comparatively expensive to pull fresh state from that.  We're still only talking about o(10) loops/

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/dynamicfile_content.go (hunk 5 above)
Comment (deads2k) [reply staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/dynamicfile_content.go]: > It looks like we cause an enqueue irrelevant of whether the bundle changes

we returned early a few lines up if nothing changed.

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticatorfactory/requestheader.go (hunk 6 above)
Comment (deads2k) [reply staging/src/k8s.io/apiserver/pkg/authentication/authenticatorfactory/requestheader.go]: > How come this needs to be duplicated? Import cycle?

Yeah, the authenticator shouldn't import an impl.

**Comment by deads2k** [comment]
Updated to the point of being able to create integration tests for reloading the certificates from disk.

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/dynamicfile_content.go [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+	if err != nil {
+		return err
+	}
+	if len(caBundle) == 0 {
+		return fmt.Errorf("missing content for CA bundle %q", c.Name())
+	}
+
+	// check to see if we have a change. If the values are the same, do nothing.
+	existing, ok := c.caBundle.Load().(*caBundleAndVerifier)
+	if ok && existing != nil && reflect.DeepEqual(existing.caBundle, caBundle) {
```
Comment (jackkleeman) [inline staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/dynamicfile_content.go]: `bytes.Equal` again here

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/dynamicfile_content.go (hunk 3 above)
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/dynamicfile_content.go]: The type conversion will panic for nil in the atomic value.

File: cmd/kubelet/app/auth.go [hunk 8]
```diff
@@ -21,10 +21,11 @@ import (
 	"fmt"
 	"reflect"
 
+	"k8s.io/apiserver/pkg/server/dynamiccertificates"
+
```
Comment (sttts) [inline cmd/kubelet/app/auth.go]: nit: order

**Comment by derekwaynecarr** [comment]
kubelet updates look good.

/approve

**Pull request merged**